### ISSUE TITLE
Fix possible preference list

### ISSUE THREAD

**Issue by cvuchener** [Dwarf-Therapist/Dwarf-Therapist#89]
From @feelotraveller's comment: https://github.com/Dwarf-Therapist/Dwarf-Therapist/issues/88#issuecomment-391224669

> Quires, scrolls and scroll-roller preferences I have seen (definitely some, pretty sure all) but they are quite rare. It was possible to add the Quire preference to roles at one stage (since my bookbinder role has it) has this since been removed? If so please give my toy back.;)

I generated a few thousands dwarves and I can confirm that incomplete tools are possible preferences (including helves). I did not see any instrument though despite the [wiki claim](http://dwarffortresswiki.org/index.php/DF2014:Preferences#Possible_Objects_of_Affection_or_Revulsion).

The good solution may be to filter *generated* items.

Creature materials need some filters too: removing generated creature (e.g. forgotten beast) materials too?

I think plants may be limited to crops.

**Comment by feelotraveller** [comment]
Good to get confirmation.

In my experience it is possible to get preferences for instruments although they are even rarer.  Seen this perhaps only once or twice since df44 released.  Not sure if this includes both multi-piece and single piece instruments...

(Additionally I specifically remember a thread a few months back from a newish player on either the gameplay or dwarf mode boards which basically went: 'WTF is an Orghal*? I've looked in the wiki and there is no mention of it.'  Attached was a screenshot of the relevant thoughts and preferences screen.  Response from another party was that it was procedurally generated instrument name, to which the reply came - 'that makes sense'.  Unable to find the thread but it does exist. *Orghal is just my makeup name for whatever the real name was.)

A bit tangential but I have never seen a preference for an instrument piece.

Barring other input I think nothing is lost by removing generated creature parts - I don't remember ever seeing preferences for them.  And matches will happen so rarely.

Plants come in a couple of different varieties - under food we want everything a dwarf could eat, including different parts like seeds (not all are edible) and leafs.  But there is also the preference for the plant itself (like there is a preference for willow wood and willow trees) that may be mininally useful occasionally.  

That said I see no problem in having a couple of categories such as 'Trees' or 'Crops' to specify for preference matches although it would be nice have the string 'Willow Trees' reported for tooltips/details.  I don't know if that would defeat the (coding) purpose?

**Comment by cvuchener** [comment]
DFHack's create-unit may not be perfect to study preferences. Dwarves created that way did not have preferences for any art form (dance, music, poems). Other preferences (like generated items) may be missing for the same reason.

> That said I see no problem in having a couple of categories such as 'Trees' or 'Crops' to specify for preference matches although it would be nice have the string 'Willow Trees' reported for tooltips/details. I don't know if that would defeat the (coding) purpose?

You mean to make the difference between Materials/Wood/willow and Trees/willows more evident? "tree" is already added in the raw name for some trees, adding "trees" after the tree name will lead to string like "apple tree trees" (similar to the current in-game description for preferences for "apple wood wood").

**Comment by feelotraveller** [comment]
Sorry I meant to test for confirmation but things have dragged, and I don't expect to get to it this weekend either.

For live plants I suspect that there might be no effect when viewed (unlike bridges).  If there was it would only be a happy thought. (Got a bit confused by the 'nature' belief.)  Need to check woodcutters and butchers (at least) though!

More generally dwarfs get happy thoughts for working with things that they have a preference for; a modifier of the 'was satisfied at work` thought.  Never noted a speed or skill gain increase.  Pretty sure that a relevant preference for a cook results in a (statistically) better quality product  - don't remember if I ever specifically tested this - mind if we are strictly talking plant preferences (as opposed to food) there won't be much if anything in this category (strawberry plants?).

Use cases can be made for all sorts of things not covered by the defaults.  Plump helmet preferences for dwarfs at the magna furnaces in the basement or for cooks producing for the the caravan?  Not my thing but both quite valid.  I would also suggest leaving open the options for colours and shapes.  (Testing for possible effects of these has not been undertaken, at least not by me, but that doesn't mean they don't exist.  Certainly imaginable that these at least might have the same effect on percieved item value as the item preferences.)

Thanks for the comments on create-unit I was thinking of using it as a shortcut but it must be somewhat wonky since, as you imply, artistic preferences are exceptionally common for in game dwarfs.

**Comment by mnorthey** [comment]
I scanned through three different saves of 222 total dwarfs, presumably _not_ randomly generated using DFHack since many of them had preferences for generated dance, music, and poetry. Here are my observations:

> In my experience it is possible to get preferences for instruments although they are even rarer.

If this is the case, then it must indeed be rare. I didn't encounter a single preference for an instrument.

> A bit tangential but I have never seen a preference for an instrument piece.

I did not encounter any preferences for instrument pieces either.

> Barring other input I think nothing is lost by removing generated creature parts - I don't remember ever seeing preferences for them. And matches would happen so rarely anyway.

Again, not a single preference for generated creature parts. In fact, the _only_ generated preferences I saw were for dance, music, and poetry. 

> I don't mind adding them (or other missing categories: color, shapes, art) if you have an use case, just give me the relevant sub-categories.

Just a warning if you do decide to these categories. I found preferences for multiple colors that were _not_ listed on the [Wiki](http://dwarffortresswiki.org/index.php/DF2014:Color#Color_tokens). They were, however, located in the DF raws file descriptor_color_standard.txt. Shapes aren't well-documented on the [Wiki](http://dwarffortresswiki.org/index.php/DF2014:Shape), which effectively loops you in circles looking for a proper list. That said, the preferences for shapes all line up with the pluralized NAME attribute in the descriptor_shape_standard.txt raw.

**Comment by mnorthey** [comment]
> But there is also the preference for the plant itself (like there is a preference for willow wood and willow trees) that may be mininally useful occasionally.

Just did some lightweight testing on this. Had a dwarf who liked cedar (the _tree_ not the wood). Put her in a variety of situations and she showed no change to mood, whether walking near a cedar, chopping it down, or hauling its wood. Game crashed before I was able to test this out with plants, but I would be surprised if it was anything different. I do recall in the past my dwarfs with strange moods depicting their preferred plants on artifacts, which I assume means the same would apply to craftsdwarfs. 

> Use cases can be made for all sorts of things not covered by the defaults. Plump helmet preferences for dwarfs at the magna furnaces in the basement or for cooks producing for the the caravan? Not my thing but both quite valid. I would also suggest leaving open the options for colours and shapes. (Testing for possible effects of these has not been undertaken, at least not by me, but that doesn't mean they don't exist. Certainly imaginable that these at least might have the same effect on percieved item value as the item preferences.)

I think per this comment there is still validity in including live plants, among other items, in preferences. Even where preferences currently have no tangible effect, there may be additional functionality in future game releases. Furthermore, I don't think we should discount the value of role-playing. I've watched enough YouTube series to know that players prioritize seemingly strange preferences for their dwarfs when it comes to roles.

**Comment by cvuchener** [comment]
Colors and shapes are read from the raws, there is no need to copy a list from the wiki. There is not much data associated with them: they are mostly words, RGB data for colors and a tile number for shape (for engravings?). Only subcategories I can imagine would be grouping by closest 4-bit color (but that depends on the palette IIRC) and shape tile.

The other missing category, food, is more complex, they have both item and material information. A quick check with 1000 dwarves gives the following item types: FISH, SEEDS, PLANT, CHEESE, PLANT_GROWTH, LIQUID_MISC, POWDER_MISC, DRINK, MEAT. These types could be used for subcategories with material based sub-subcategories? Drinks: animal, plants. Liquids: milk, honey, oil. Plant growths: fruits, leaves. There is still to find the possible materials.

**Comment by feelotraveller** [comment]
>If this is the case, then it must indeed be rare. I didn't encounter a single preference for an instrument.

Yes, rare like maybe one in a thousand, or one in ten thousand.  The wiki also lists instruments as possible preferences.  (And I wish I could find that thread again...)

If you didn't find anything supporting happy thought effects of live plant preferences in testing I could well be wrong on that.  But I too don't see any harm in including them.

>The other missing category, food, is more complex

As far as food goes we definitely want to include everything that can be eaten by dwarfs that is a possible preference.  In addition to the above there are also the eggs and various kinds of press cakes that I have seen preferences for.  I don't think that any of the Misc. Liquids can be consumed.  Not sure what the Misc. Powders contains but if it's equivalent to Milled Plants in game then most can be eaten but not the dyes.  Some of the extracts can be consumed but not others, likewise with seeds.  There are also (I think?) food preferences which cannot be satisfied in the game, for example Bumblebee Mead cannot be made/acquired (except perhaps with DFHack).  

When in doubt I think it is better to err on the side of caution and include possibly useless preferences.

**Comment by cvuchener** [comment]
More tests on 2000 dwarves (with create-unit):

>  In addition to the above there are also the eggs and various kinds of press cakes that I have seen preferences for.

Only egg I see is "feather tree frozen egg yolk" and I think it is a PLANT_GROWTH. Maybe another limitation from create-unit.

>  I don't think that any of the Misc. Liquids can be consumed. 

But they can be included in prepared meals, right? I think that still count.

> Not sure what the Misc. Powders contains but if it's equivalent to Milled Plants in game then most can be eaten but not the dyes. 

All the materials I found in POWDER_MISC food preferences have the EDIBLE_COOKED flag (maybe EDIBLE_RAW is enough too).

**Comment by cvuchener** [comment]
One remark about meat preferences, the game (and currently DT) only gives the creature name, but the preference actually references a specific body part material.

I tried with a dwarf that "prefers to consume sea lamprey" (actually sea lamprey eye), I made him eat sea lamprey liver, it did not fulfill his "good meal" need, but eating sea lamprey eye did. The body part is important.

When food preferences are properly implemented in DT, I think it should show the exact body part unlike DF.

**Comment by cvuchener** [comment]
I added a few of the requested changes in #102:
 - removed generated items and re-added incomplete ones
 - added colors and shapes

With a full list of all shapes, it became apparent that they were missing the adjectives. The in-game description is also missing the adjectives. 

> likes hematites, pig iron, green diamond, gems, bucklers, catapult parts, giant peach-faced lovebirds, ...

It is hard to guess that "gems" actullay means "pear cut gems shape". But some shape have multiple adjectives I don't know how to choose (or not choose).

**Comment by cvuchener** [comment]
I think PLANT food preferences only uses the basic material (BASIC_MAT in the raws). Unlike meat, plant food preferences always reference the plant itself and not any of its growths/products.

@feelotraveller, since you wanted to discuss preference names, check [that dfhack script](https://pastebin.com/PmkPyGe1) and tell me if you are happy with the names it prints. There is a few issues (see comments), but I think the food preference names are good.

**Comment by cvuchener** [comment]
If your are not happy with "long yams plants" you will have to edit your raws. I won't do anything too complex with names, only concatenating strings that exists in the raws and maybe fixed ones (e.g. "the sight of" for dances). Same for other plants and trees, a lot of your suggestions should go to Toady, not me.

> This comes from the capitalised part of the script and not what I take to be the output.

The capitalized part is the internal data so you know the exact preference, the next string is the user-friendly string that I try to build from raw data. I should have been clearer about that.

Decoration should added around that string: "the color" for color preferences, "the shape of" for shape (currently in DT but not in-game IIRC), "the words of" for poetry, "the sight of" for dances, "fabric" for THREAD_PLANT+SOLID materials, "wood" for WOOD materials (you can see them in [preference.cpp](https://github.com/Dwarf-Therapist/Dwarf-Therapist/blob/master/src/preference.cpp) get_description functions). "wood" leads to repetitions, but since DF also have them, I left it. Food/Plant/Creature confusion is avoided by having the food in a different sentence ("prefers to consume ...").

For shapes I missed that there were flags. I could use them to decide if the adjectives or names should be used.

```
[SHAPE:STAR]
	[NAME:star:stars]
	[WORD:STAR]
	[ADJ:three-pointed]
	[ADJ:four-pointed]
	[ADJ:five-pointed]
	[ADJ:many-pointed]
	[TILE:'*']
```
*Urist likes (the shape of) stars.* (default to name only if there is no flags)

```
[SHAPE:GIZZARD_STONE]
	[NAME:gizzard stone:gizzard stones]
	[GEMS_USE_NOUN]
	[TILE:9]
```
*Urist likes (the shape of) gizzard stones.*

```
[SHAPE:SMOOTH_PEBBLE]
	[NAME:pebble:pebbles]
	[ADJ:smooth]
	[GEMS_USE_ADJ_NOUN]
	[TILE:9]
```
*Urist likes (the shape of) smooth pebbles.*

```
[SHAPE:POINT_CUT_GEM]
	[NAME:gem:gems]
	[ADJ:point cut]
	[GEMS_USE_ADJ]
	[TILE:4]
```
*Urist likes (the shape of) point cut.* Or force the use of "gems" in this context. *Urist likes (the shape of) point cut gems.*"

Shapes with GEMS_USE_ADJ* all have exactly one adjective, so choosing is not an issue.

Having a "gem cut" sub-category for shapes would be interesting, but it may be difficult to implement.

**Comment by feelotraveller** [comment]
>check that dfhack script and tell me if you are happy with the names it prints

And that's what I did. :stuck_out_tongue:  What you do with that is your business.  Thanks for the opportunity to provide feedback.

Like your suggestions for the gems (but the cut category would be useful...) so I am happy that something I said led to that.

All the best.
